Drupal 7.39 breaks autocomplete functionality

slangerx opened the pull request:
<a href="http://drupal.org/drupal-7.39-release-notes">Drupal 7.39</a> came out on August 19, 2015 -- a security release. One of the vulnerabilities it was patching was <em>autocomplete</em>, which was open to cross-site scripting (<a href="http://drupal.org/SA-CORE-2015-003">SA-CORE-2015-003</a>). A new function was introduced (<a href="https://github.com/drupal/drupal/commit/be00a1ced4104d84df2f34b149b35fb0adf91093#diff-76b8e4a4630d8b8b09a8579c35fd31e9R3965">form_process_autocomplete</a>) to "sufficiently sanitize" requested URLs. Autocomplete broke for many modules after upgrading to Drupal 7.39 -- including islandora_xml_forms. This is because the function form_process_autocomplete() is now required as part of an element's "#process" array.

For islandora_xml_forms, this problem only appears to impact autocomplete fields where multiple values can be added. Single-item autocomplete fields seem unaffected by the upgrade.

Autocomplete (under Drupal 7.39) also seems to have a problem with field IDs with numbers in them. The module islandora_xml_forms has been overriding the element ID (which is usually based on the field name: <em>example:</em> <tt>formSections[subjects][topic]</tt> becomes <em>edit-formsections-subjects-topic</em>) and replacing it with a hash instead (<em>00000000682de34e000000015b767354</em>). As of 7.39, numbers in an ID seem to break the autocomplete functionality, which is why I reinstated the Drupal-built IDs. Multi-item fields can sometimes have zeroes at the end of their IDs, which is why numbers are being stripped out.

slangerx committed:
Autocomplete fix related to Drupal 7.39.

slangerx committed:
Make autocomplete fix for Drupal 7.39 backward-compatible.

slangerx commented:
@ruebot Sure! Done: https://jira.duraspace.org/browse/ISLANDORA-1414

adam-vessey commented:
I'm... uncomfortable with the ID changes... Do we know the exact code in Drupal which changed to mess with numerical IDs? Not 100% sure, but I think there is some code which depends on the `#id` bit matching `#hash`, which this would break.

DiegoPino commented:
@adam-vessey is right. I already added a comment on this in the JIRA ticket. Let me do some test this weekend before going with this

slangerx commented:
@adam-vessey @DiegoPino  I think I know what the problem is. It all comes down to a single line -- the <a href="https://github.com/drupal/drupal/commit/be00a1ced4104d84df2f34b149b35fb0adf91093#diff-76b8e4a4630d8b8b09a8579c35fd31e9L3937"><tt>$attributes['id']</tt> that's being set here</a>. 

If you upgrade to Drupal 7.39; patch <tt>xml_form_elements</tt> with my proposed <tt>hook_element_info_alter()</tt> change; leave the element IDs alone (so #id = #hash, just as before); and then switch that one line (<tt>$attributes['id'] = ...</tt>)  back to the Drupal 7.38 version, autocomplete works just fine. If you switch that line back to the 7.39 version, it breaks again. 

Why? Because the hash seems to be a moving target. It appears that when the element gets processed a second time (<tt>xml_form_elements_tag_process</tt>, <tt>form_process_autocomplete</tt>), the element hash changes and so does the element ID -- at least from <tt>form_process_autocomplete</tt>'s point-of-view. In the end, <tt>$element['#id']</tt> and <tt>$element['#attributes']['id']</tt> match in the element array -- two values used by previous versions of Drupal. However, <tt>$element['#autocomplete_input']['#id']</tt> -- now used by Drupal 7.39 -- has a slightly different hash ID than the <tt>$element['#id']</tt>. With mismatched IDs, the autocomplete functionality can't attach to the field properly and the process breaks.

If there was some way to keep the hash value (and subsequently, the element IDs) stable, I think this problem would be solved.

DiegoPino commented:
Thanks @slangerx, i will test some options(less code one will win) here and tell you about this on monday.

DiegoPino commented:
Ok, found the problem @slangerx and @adam-vessey.
The hidden autocomplete container get's the id + '-autocomplete' of the tag. It's ok, because the autocomplete path is assigned to the tag (the "0" element). But the textfield gets the 'id' of the wrapping 'tags' elements. So no match and javascript error when running through autocomplete.js.
It's induced by the changes in 7.39 but looks like it's our error/implementation at the end? Will try to find a simple fix that does not involve hacking more than necessary and not touching the hashes.

ruebot committed:
Merge pull request #196 from manez/7.x

ISLANDORA-1396 and ISLANDORA-1394

DiegoPino commented:
Found a fix, still have to test on previous drupal and using pre filled forms with templates: 

There is no real reason to swap id's on https://github.com/slangerx/islandora_xml_forms/commit/e7cf988cc5adb23a9b42f99cf0b9cabe3f924e35#diff-c67eeb88200fa57a2bd06a0bf4167f28L228 
Currently on template_preprocess_tags we assign our textfield the id of the first found tag. It's not a good idea to change id's at theming level. All process function already passed so we just confuse Drupal. Keeping the id assigned at the beginning works. Also makes sense in general.

Will do some extra testing and if everything goes well i will pull against you @slangerx so you can review it using your own forms.

slangerx committed:
Restore Tag.inc & Tags.inc to previous version. Let id=hash.

slangerx commented:
@DiegoPino Sounds good! I've updated my pull request, restoring "Tag.inc" and "Tags.inc" to their original versions. Once again, <tt>$element['#id'] = $element['#hash'];</tt>.

DiegoPino commented:
Hi @slangerx, i just did a pull request https://github.com/slangerx/islandora_xml_forms/pull/1
You have travis also, but it's failing (not my fault!  :innocent: ) Since you did not have a ISLANDORA-1404 branch and you already made this pull request i pulled against your 7.x. Also your Fork was not in sync with islandora's head, so the pull request also includes those extras to put you up-to-date. Just merge(should appear here by "magic") and test.
Tell me if you find any errors or problems!

DiegoPino committed:
Merge remote-tracking branch 'slangerx/7.x' into ISLANDORA-1414

DiegoPino committed:
ISLANDORA-1414

Ultra simple fix for Drupal 7.39 Autocomplete functionality on tags
based on @slangerx identification and previous fix. We no longer
reassign element #id for textfield in the
template_preprocess_tags(&$vars) , we use the same hash assigned
initially. Tested on  Drupal 7.37-7.39 with blank form and form using
templates pre filled for tags.

slangerx committed:
Merge pull request #1 from DiegoPino/ISLANDORA-1414

Islandora 1414

Commit:
Correct comment formatting. Cause of Travis CI fail?

DiegoPino commented:
@slangerx, hi. Can you confirm you have tags, tag and autocomplete working as supposed? Thanks!

slangerx commented:
@DiegoPino I just tested our combined changes and from my dev site's perspective, the problem is fixed. Both multi-item and single-item autocomplete fields are working again. Thanks! P.S. A pull request of a pull request? I didn't know you could do that! Very clever. :-)

DiegoPino commented:
Hi, thanks @slangerx , last test sorry. Can you confirm the resulting XML looks good?

slangerx commented:
@DiegoPino I just ran another test in our dev environment. On the interface side, the autocomplete fields worked correctly. The resulting XML MODS looks exactly the way I expect it to. The error logs are clean. :-)

DiegoPino commented:
By the way: Pull of a pull was @whikloj idea. Nice right? Thanks for the testing.

DiegoPino commented:
Merge :thumbsup:

slangerx commented:
:thumbsup::thumbsup:
